lucetc/src/compiler.rs

```diff
@@ -47,6 +47,91 @@ impl OptLevel {
     }
 }
 
+pub struct CompilerBuilder {
+    pub target: Triple,
+    opt_level: OptLevel,
+    cpu_features: CpuFeatures,
+    heap_settings: HeapSettings,
+    count_instructions: bool,
+    canonicalize_nans: bool,
+    validator: Option<Validator>,
+}
+
+impl CompilerBuilder {
+    pub fn new() -> Self {
+        Self {
+            target: Triple::host(),
+            opt_level: OptLevel::default(),
+            cpu_features: CpuFeatures::default(),
+            heap_settings: HeapSettings::default(),
+            count_instructions: false,
+            canonicalize_nans: false,
+            validator: None,
+        }
+    }
+
+    pub fn target(&mut self, target: Triple) -> &mut Self {
+        self.target = target;
+        self
+    }
+
+    pub fn opt_level(&mut self, opt_level: OptLevel) -> &mut Self {
+        self.opt_level = opt_level;
+        self
+    }
+
+    pub fn cpu_features(&mut self, cpu_features: CpuFeatures) -> &mut Self {
+        self.cpu_features = cpu_features;
+        self
+    }
+
+    pub fn cpu_features_mut(&mut self) -> &mut CpuFeatures {
+        &mut self.cpu_features
+    }
```

Comment:
Building off of the `with_` changes, I would change the signature of these mutator methods to look like

```rust
cpu_features(&mut self, cpu_features: CpuFeatures)
```

This means code that changes the value of a member, `CpuFeatures` in this case, does so by passing the new value _to_ our builder rather than borrowing and mutating the existing value.

Reply:
so if you look at the example of how we expose `heap_settings` to enable `min_reserved_size`, `LucetOpts` wants to be able to manipulate _parts_ of `heap_settings`, not wholly overwrite it, which is what a signature like that would provide.  Alternately, I could not expose `heap_settings` at all (like `LucetOpts` doesn't) and instead expose each of the individual modifiable parts (although that is opposite of the pattern it uses with `cpu_features`).  I was trying to find a consistent way to handle both `cpu_features` and `heap_settings` where they don't seem to be handled consistently in `LucetOpts`.

Reply:
Ah! Thank you for noticing this, it is very thoughtful.

I think the way you have this currently is great! Sometimes exceptions are needed, so an additional `cpu_features_mut` method because of implementation details seems perfectly reasonable to me. The  `min_reserved_size` methods feel more like concerns of the public API, so I don't think this struct needs to mirror that.

Reply:
I commented elsewhere about this, but the `target_ref` method is another case like this. I think it's fine to have these, and the rest of the methods are there to provide a nice API surface for general-use.